src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extensions.cs

```diff
@@ -0,0 +1,539 @@
+                    else if (partUnderlyingType is not null
+                        && partUnderlyingType.TypeKind != TypeKind.Error
+                        && !TypeSymbol.Equals(partUnderlyingType, underlyingType, TypeCompareKind.ConsiderEverything))
+                    {
+                        // the parts do not agree
+                        bool shouldReportUnderlyingTypeConflict = false;
+                        if (partUnderlyingType.Equals(underlyingType, TypeCompareKind.ObliviousNullableModifierMatchesAny))
+                        {
+                            if (ContainsOnlyOblivious(underlyingType))
+                            {
+                                underlyingType = partUnderlyingType;
+                                underlyingTypeLocation = declaration.NameLocation;
+                            }
+                            else if (!ContainsOnlyOblivious(partUnderlyingType))
+                            {
+                                shouldReportUnderlyingTypeConflict = true;
+                            }
+                        }
+                        else
+                        {
+                            shouldReportUnderlyingTypeConflict = true;
+                        }
+
+                        if (shouldReportUnderlyingTypeConflict)
+                        {
+                            var info = diagnostics.Add(ErrorCode.ERR_PartialMultipleUnderlyingTypes, Locations.FirstOrNone(), this);
+                            underlyingType = new ExtendedErrorTypeSymbol(underlyingType, LookupResultKind.Ambiguous, info);
+                            underlyingTypeLocation = declaration.NameLocation;
+                            reportedUnderlyingTypeConflict = true;
+                        }
+                    }
+                }
+
+                foreach (NamedTypeSymbol partBaseExtension in one.BaseExtensions)
+                {
+                    if (!baseExtensionLocations.ContainsKey(partBaseExtension))
+                    {
+                        baseExtensionsBuilder.Add(partBaseExtension);
+                        baseExtensionLocations.Add(partBaseExtension, declaration.NameLocation);
+                    }
+                }
+            }
+
+            var useSiteInfo = new CompoundUseSiteInfo<AssemblySymbol>(diagnostics, ContainingAssembly);
+
+            if (!sawUnderlyingType)
+            {
+                diagnostics.Add(ErrorCode.ERR_ExtensionMissingUnderlyingType, Locations.FirstOrNone(), this);
+            }
+            else if (underlyingType is not null)
+            {
+                Debug.Assert(underlyingTypeLocation != null);
+
+                if (!this.IsNoMoreVisibleThan(underlyingType, ref useSiteInfo))
+                {
+                    diagnostics.Add(ErrorCode.ERR_BadVisUnderlyingType, underlyingTypeLocation, this, underlyingType);
+                }
+
+                if (underlyingType.HasFileLocalTypes() && !this.HasFileLocalTypes())
```

Comment:
>this.HasFileLocalTypes() [](http://example.com/codeflow?start=59&length=24)

Is this supposed to check in this is ```IsFileLocal```. Consider doing that instead, for clarity.